# Zillow Rental Data Set
## KEY: ZillowRent1
## Source: 

Zillow Observed Rent Index (ZORI) Smoothed, Seasonally Adjusted. (2014–2020, January 1–August 31). [A smoothed measure of the typical observed market rate rent across a given region.]. Zillow Research. https://www.zillow.com/research/data/

In [2]:
# import needed libraries
from numpy import *
import pandas as pd
import matplotlib.pyplot as plt

pd.set_option('display.max_rows', None)

pd.set_option('display.max_columns', None)

In [3]:
# load the zillow rental csv into a dataframe
dfZRental = pd.read_csv(r'../DataSet/Zillow_Rental_2014-2020.csv',header=0,encoding = "ISO-8859-1")

In [4]:
# show the first 5 rows to check data loaded as epected
dfZRental.head(5)

RegionID                          RegionName  SizeRank  2014-01  2014-02  \
0    102001                       United States         0   1349.0     1354   
1    394913                        New York, NY         1   2318.0     2327   
2    753899  Los Angeles-Long Beach-Anaheim, CA         2   1856.0     1869   
3    394463                         Chicago, IL         3   1533.0     1537   
4    394514               Dallas-Fort Worth, TX         4   1221.0     1227   

   2014-03  2014-04  2014-05  2014-06  2014-07  2014-08  2014-09  2014-10  \
0   1360.0     1365     1371     1376     1381     1387     1392     1398   
1   2336.0     2345     2354     2363     2372     2381     2390     2399   
2   1882.0     1894     1907     1920     1932     1945     1957     1970   
3   1541.0     1545     1548     1552     1556     1559     1563     1567   
4   1233.0     1239     1245     1251     1257     1262     1268     1274   

   2014-11  2014-12  2015-01  2015-02  2015-03  2015-04  2015-05  2015-06  \
0     1403     1409     1414     1419     1425     1430     1436     1441   
1     2408     2417     2426     2435     2443     2452     2461     2469   
2     1982     1994     2006     2017     2028     2040     2052     2064   
3     1570     1574     1577     1582     1586     1590     1595     1599   
4     1280     1287     1293     1298     1304     1310     1315     1321   

   2015-07  2015-08  2015-09  2015-10  2015-11  2015-12  2016-01  2016-02  \
0     1446     1452     1457     1462     1467     1472     1477     1482   
1     2477     2485     2492     2499     2504     2510     2515     2520   
2     2075     2088     2100     2112     2124     2136     2147     2158   
3     1604     1609     1615     1620     1625     1630     1636     1640   
4     1326     1331     1336     1341     1346     1352     1357     1362   

   2016-03  2016-04  2016-05  2016-06  2016-07  2016-08  2016-09  2016-10  \
0     1486     1491     1495     1499     1504     1507     1511     1515   
1     2524     2528     2532     2536     2539     2543     2546     2549   
2     2170     2181     2191     2201     2212     2221     2231     2241   
3     1644     1648     1651     1654     1658     1661     1663     1666   
4     1368     1374     1379     1385     1391     1396     1402     1407   

   2016-11  2016-12  2017-01  2017-02  2017-03  2017-04  2017-05  2017-06  \
0     1519     1523     1527     1530     1534     1537     1541     1544   
1     2552     2555     2559     2561     2564     2567     2569     2571   
2     2250     2260     2270     2279     2288     2297     2305     2314   
3     1669     1672     1675     1677     1680     1682     1684     1687   
4     1412     1417     1422     1426     1431     1435     1439     1443   

   2017-07  2017-08  2017-09  2017-10  2017-11  2017-12  2018-01  2018-02  \
0     1548     1552     1556     1559     1564     1568     1572     1577   
1     2573     2575     2576     2578     2580     2582     2584     2587   
2     2322     2330     2338     2346     2353     2361     2368     2376   
3     1689     1691     1694     1696     1698     1700     1703     1705   
4     1447     1451     1455     1459     1462     1465     1469     1472   

   2018-03  2018-04  2018-05  2018-06  2018-07  2018-08  2018-09  2018-10  \
0     1581     1586     1591     1595     1600     1605     1609     1614   
1     2590     2593     2597     2600     2604     2608     2612     2616   
2     2384     2392     2401     2409     2417     2425     2432     2440   
3     1707     1710     1712     1715     1717     1720     1723     1726   
4     1476     1479     1483     1486     1490     1494     1498     1502   

   2018-11  2018-12  2019-01  2019-02  2019-03  2019-04  2019-05  2019-06  \
0     1619     1623     1628     1632     1637     1641     1646     1651   
1     2621     2625     2630     2635     2641     2646     2652     2658   
2     2447     2454     2461     2467     2474     2480 

In [5]:
# show last 5 rows to check data loaded as epected
dfZRental.tail(5)

RegionID          RegionName  SizeRank  2014-01  2014-02  2014-03  \
101    394995  Port St. Lucie, FL       119   1155.0     1163   1171.0   
102    394602    Fort Collins, CO       159   1250.0     1262   1274.0   
103    394405         Boulder, CO       162   1531.0     1541   1551.0   
104    394645         Greeley, CO       177   1285.0     1294   1303.0   
105    394623     Gainesville, GA       229    984.0      988    993.0   

     2014-04  2014-05  2014-06  2014-07  2014-08  2014-09  2014-10  2014-11  \
101     1179     1186     1194     1202     1209     1217     1224     1231   
102     1286     1297     1309     1321     1332     1343     1355     1364   
103     1562     1572     1582     1592     1603     1614     1625     1636   
104     1312     1320     1329     1338     1346     1355     1364     1373   
105      997     1002     1007     1012     1017     1022     1026     1031   

     2014-12  2015-01  2015-02  2015-03  2015-04  2015-05  2015-06  2015-07  \
101     1237     1244     1250     1256     1261     1268     1274     1280   
102     1374     1384     1392     1400     1408     1413     1419     1425   
103     1647     1658     1670     1682     1694     1708     1722     1736   
104     1382     1391     1398     1406     1414     1422     1429     1437   
105     1036     1041     1047     1053     1059     1066     1072     1078   

     2015-08  2015-09  2015-10  2015-11  2015-12  2016-01  2016-02  2016-03  \
101     1288     1296     1304     1314     1324     1333     1342     1351   
102     1429     1434     1439     1444     1449     1454     1458     1462   
103     1747     1759     1770     1779     1787     1796     1800     1805   
104     1443     1449     1456     1462     1468     1473     1478     1483   
105     1083     1087     1092     1098     1103     1109     1115     1121   

     2016-04  2016-05  2016-06  2016-07  2016-08  2016-09  2016-10  2016-11  \
101     1360     1366     1373     1379     1384     1390     1395     1400   
102     1467     1470     1474     1477     1481     1484     1487     1489   
103     1810     1812     1813     1815     1817     1818     1820     1822   
104     1488     1491     1494     1498     1501     1504     1507     1511   
105     1128     1133     1139     1145     1151     1157     1163     1169   

     2016-12  2017-01  2017-02  2017-03  2017-04  2017-05  2017-06  2017-07  \
101     1405     1410     1416     1421     1426     1431     1437     1443   
102     1491     1493     1495     1497     1499     1502     1505     1508   
103     1825     1828     1832     1836     1840     1846     1851     1857   
104     1514     1518     1523     1529     1534     1540     1547     1553   
105     1176     1182     1188     1195     1201     1207     1213     1218   

     2017-08  2017-09  2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  \
101     1450     1457     1465     1473     1480     1488     1497     1505   
102     1511     1514     1517     1521     1525     1529     1534     1539   
103     1863     1868     1874     1879     1885     1890     1896     1902   
104     1560     1567     1573     1580     1586     1593     1598     1603   
105     1223     1228     1234     1238     1242     1246     1250     1254   

     2018-04  2018-05  2018-06  2018-07  2018-08  2018-09  2018-10  2018-11  \
101     1513     1521     1529     1537     1542     1547     1553     1556   
102     1543     1548     1553     1557     1562     1567     1572     1576   
103     1908     1912     1917     1921     1926     1930     1935     1940   
104     1609     1614     1620     1625     1632     1638     1645     1650   
105     1258     1265     1271     1277     1285     1293     1300     1307   

     2018-12  2019-01  2019-02  2019-03  2019-04  2019-05  2019-06  2019-07  \
101     1560     1564     1568     1572     1576     1581     1586     1592   
102     1580     1585     1589     1592     1596     1600     1604     1608  

In [6]:
# run describe to get basics
dfZRental.describe(include="all")

RegionID          RegionName    SizeRank      2014-01  \
count      106.000000                 106  106.000000    99.000000   
unique            NaN                 106         NaN          NaN   
top               NaN  Port St. Lucie, FL         NaN          NaN   
freq              NaN                   1         NaN          NaN   
mean    402173.283019                 NaN   55.933962  1180.444444   
std      66680.198177                 NaN   39.272981   332.744071   
min     102001.000000                 NaN    0.000000   601.000000   
25%     394497.500000                 NaN   26.250000   975.500000   
50%     394802.500000                 NaN   52.500000  1103.000000   
75%     395029.500000                 NaN   79.750000  1256.500000   
max     753924.000000                 NaN  229.000000  2318.000000   

            2014-02      2014-03      2014-04      2014-05      2014-06  \
count    106.000000   104.000000   106.000000   106.000000   106.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean    1171.113208  1183.692308  1181.301887  1186.264151  1191.330189   
std      330.190517   332.499539   337.511966   341.213110   344.868497   
min      602.000000   602.000000   603.000000   603.000000   604.000000   
25%      968.000000   979.500000   973.750000   977.500000   981.750000   
50%     1093.000000  1100.000000  1101.500000  1105.000000  1109.000000   
75%     1259.000000  1270.250000  1269.000000  1272.250000  1275.750000   
max     2327.000000  2336.000000  2345.000000  2354.000000  2363.000000   

            2014-07      2014-08      2014-09      2014-10      2014-11  \
count    106.000000   106.000000   106.000000   106.000000   106.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean    1196.301887  1201.292453  1206.367925  1211.377358  1216.160377   
std      348.617902   352.426689   356.189632   360.056886   363.519545   
min      605.000000   605.000000   605.000000   606.000000   607.000000   
25%      986.250000   988.750000   990.250000   991.750000   994.750000   
50%     1112.000000  1115.500000  1119.000000  1122.500000  1126.000000   
75%     1279.250000  1282.750000  1286.250000  1289.750000  1293.000000   
max     2376.000000  2408.000000  2441.000000  2473.000000  2504.000000   

            2014-12      2015-01      2015-02      2015-03      2015-04  \
count    106.000000   106.000000   106.000000   106.000000   106.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean    1220.830189  1225.622642  1230.169811  1234.698113  1239.235849   
std      367.070552   370.570974   373.643947   376.682996   379.798808   
min      607.000000   608.000000   609.000000   611.000000   612.000000   
25%      998.250000  1001.750000  1004.000000  1006.250000  1005.500000   
50%     1129.000000  1132.000000  1136.000000  1139.500000  1143.500000   
75%     1299.750000  1306.000000  1314.500000  1320.750000  1327.500000   
max     2535.000000  2566.000000  2594.000000  2622.000000  2650.000000   

            2015-05      2015-06      2015-07      2015-08      2015-09  \
count    106.000000   106.000000   106.000000   106.000000   106.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean    1243.735849  1248.301887  1252.877358  1257.292453  1261.820755   
std      382.625101

**Notice** 2014-01 and 2014-03 are missing some values

There are **106 unique region names**, what one are in Virginia


In [7]:
 dfZRental[dfZRental['RegionName'].str.contains("VA")]

RegionID          RegionName  SizeRank  2014-01  2014-02  2014-03  \
36    395194  Virginia Beach, VA        36   1204.0     1205   1207.0   
45    395022        Richmond, VA        45      NaN     1092   1093.0   

    2014-04  2014-05  2014-06  2014-07  2014-08  2014-09  2014-10  2014-11  \
36     1209     1210     1212     1213     1215     1216     1218     1219   
45     1094     1096     1097     1098     1100     1101     1102     1104   

    2014-12  2015-01  2015-02  2015-03  2015-04  2015-05  2015-06  2015-07  \
36     1220     1221     1223     1224     1225     1226     1228     1229   
45     1105     1107     1109     1111     1114     1116     1119     1122   

    2015-08  2015-09  2015-10  2015-11  2015-12  2016-01  2016-02  2016-03  \
36     1231     1232     1234     1235     1237     1238     1239     1240   
45     1125     1127     1130     1133     1137     1140     1143     1147   

    2016-04  2016-05  2016-06  2016-07  2016-08  2016-09  2016-10  2016-11  \
36     1241     1242     1243     1243     1244     1245     1246     1247   
45     1150     1153     1156     1158     1160     1163     1165     1167   

    2016-12  2017-01  2017-02  2017-03  2017-04  2017-05  2017-06  2017-07  \
36     1249     1250     1252     1253     1255     1257     1259     1260   
45     1169     1172     1174     1176     1179     1181     1185     1188   

    2017-08  2017-09  2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  \
36     1262     1264     1265     1267     1269     1271     1273     1275   
45     1192     1196     1200     1204     1208     1212     1216     1220   

    2018-04  2018-05  2018-06  2018-07  2018-08  2018-09  2018-10  2018-11  \
36     1277     1279     1282     1284     1287     1289     1292     1295   
45     1224     1228     1232     1235     1239     1242     1245     1249   

    2018-12  2019-01  2019-02  2019-03  2019-04  2019-05  2019-06  2019-07  \
36     1298     1301     1304     1307     1310     1314     1317     1320   
45     1252     1255     1259     1262     1265     1269     1272     1275   

    2019-08  2019-09  2019-10  2019-11  2019-12  2020-01  2020-02  2020-03  \
36     1322     1325     1328     1330     1333     1335     1338     1341   
45     1279     1282     1285     1289     1292     1295     1298     1301   

    2020-04  2020-05  2020-06  2020-07  2020-08  
36     1344     1346     1349     1352     1354  
45     1304     1307     1310     1313     1316

What about **NY state**?

In [8]:
 dfZRental[dfZRental['RegionName'].str.contains(", NY")]

RegionID     RegionName  SizeRank  2014-01  2014-02  2014-03  2014-04  \
1     394913   New York, NY         1   2318.0     2327   2336.0     2345   
47    394425    Buffalo, NY        47    844.0      845    846.0      847   
51    395031  Rochester, NY        51    902.0      905    908.0      911   
60    394308     Albany, NY        60   1129.0     1131   1134.0     1137   
78    395143   Syracuse, NY        79    861.0      865    868.0      871   

    2014-05  2014-06  2014-07  2014-08  2014-09  2014-10  2014-11  2014-12  \
1      2354     2363     2372     2381     2390     2399     2408     2417   
47      848      850      851      852      854      855      858      860   
51      913      916      919      921      924      926      929      931   
60     1139     1141     1143     1146     1148     1150     1152     1154   
78      874      877      880      883      886      888      891      893   

    2015-01  2015-02  2015-03  2015-04  2015-05  2015-06  2015-07  2015-08  \
1      2426     2435     2443     2452     2461     2469     2477     2485   
47      862      866      870      874      878      882      887      890   
51      933      935      937      940      942      945      947      950   
60     1155     1155     1155     1156     1157     1158     1159     1162   
78      895      896      897      898      901      903      905      909   

    2015-09  2015-10  2015-11  2015-12  2016-01  2016-02  2016-03  2016-04  \
1      2492     2499     2504     2510     2515     2520     2524     2528   
47      894      898      901      904      907      910      913      916   
51      953      956      959      962      965      967      970      972   
60     1165     1167     1170     1173     1176     1179     1182     1185   
78      912      916      919      922      925      926      928      929   

    2016-05  2016-06  2016-07  2016-08  2016-09  2016-10  2016-11  2016-12  \
1      2532     2536     2539     2543     2546     2549     2552     2555   
47      919      923      926      930      933      937      940      943   
51      974      976      978      980      982      984      986      988   
60     1188     1192     1196     1199     1203     1206     1209     1212   
78      930      931      933      935      937      939      942      944   

    2017-01  2017-02  2017-03  2017-04  2017-05  2017-06  2017-07  2017-08  \
1      2559     2561     2564     2567     2569     2571     2573     2575   
47      946      949      951      954      956      957      959      961   
51      990      992      994      996      998     1000     1002     1005   
60     1215     1218     1220     1223     1224     1225     1225     1227   
78      947      951      955      958      962      966      970      973   

    2017-09  2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  \
1      2576     2578     2580     2582     2584     2587     2590     2593   
47      963      964      966      968      970      972      975      977   
51     1008     1010     1013     1016     1018     1021     1024     1027   
60     1228     1229     1232     1235     1238     1242     1246     1250   
78      976      979      982      985      988      991      993      995   

    2018-05  2018-06  2018-07  2018-08  2018-09  2018-10  2018-11  2018-12  \
1      2597     2600     2604     2608     2612     2616     2621     2625   
47      980      983      985      989      992      995      998     1002   
51     1029     1032     1034     1036     1038     1041     1043     1045   
60     1254     1258     1262     1264     1266     1268     1270     1271   
78      997      998      999      999      999      999      999     1000   

    2019-01  2019-02  2019-03  2019-04  2019-05  2019-06  2019-07  2019-08  \
1      2630     2635     2641     2646     2652     2658     2664     2669   
47     1005     1008     1012     1015     1018     1020     1023     1026   
51     1047     1049    

#### Pivot the table on dates in order to get a time series


In [9]:
#use melt to change monthly dates into rows
dfZRentalPivot = dfZRental.melt(id_vars=['RegionID','RegionName', 'SizeRank'], var_name='YYYY-MM-DD')

dfZRentalPivot['yr'] = pd.DatetimeIndex(dfZRentalPivot['YYYY-MM-DD']).year
dfZRentalPivot['period'] = pd.DatetimeIndex(dfZRentalPivot['YYYY-MM-DD']).month

dfZRentalPivot.head(5)


RegionID                          RegionName  SizeRank YYYY-MM-DD   value  \
0    102001                       United States         0    2014-01  1349.0   
1    394913                        New York, NY         1    2014-01  2318.0   
2    753899  Los Angeles-Long Beach-Anaheim, CA         2    2014-01  1856.0   
3    394463                         Chicago, IL         3    2014-01  1533.0   
4    394514               Dallas-Fort Worth, TX         4    2014-01  1221.0   

     yr  period  
0  2014       1  
1  2014       1  
2  2014       1  
3  2014       1  
4  2014       1

In [10]:
dfZRentalPivot.dtypes

RegionID        int64
RegionName     object
SizeRank        int64
YYYY-MM-DD     object
value         float64
yr              int64
period          int64
dtype: object

### Basic Dataframe Information
1. Will need to massage data
    1. pivot - **DONE**
    1. standardize date format - **DONE**
    1. indicate monthly period - **DONE**
    1. possibly create a quarterly mean for q1,q2,q3,q4 to match other data sets
    1. possibly create an annual mean for each city to match other data sets
1. Almost no missing data
1. Two Virginia cities are represented, including Virginia Beach, and Richmond
1. Date Range
    1. 01/2014 - 08/2020
1. Time Scale
    1. Monthly
1. Geo scale
    1. Locality
1. Data is smoothed seasonally adjusted rent $

### Next Steps
1. Get the average rent by year for the two Virginia cities.
1. Apply inflation method to put dollars in base year 2019 dollars.
1. Add GDP into the DataFrame.
1. Plot Average Rent Against GDP and explain conclusion.

**Nice to have:** Quartely GDP data to have 28 data points for comparison rather than just 7.